src/filesystem/impls/filer/UrlCache.js

```diff
@@ -0,0 +1,381 @@
+                         "Some preview features may not work correctly (e.g., dynamic script loading). " +
+                         "Consider using a browser that does, see: "                                     +
+                         "https://jakearchibald.github.io/isserviceworkerready");
+        }
+
+    }
+    BlobUrlProvider.prototype = Object.create(UrlProviderBase.prototype);
+    BlobUrlProvider.prototype.constructor = BlobUrlProvider;
+
+    BlobUrlProvider.prototype.init = function(callback) {
+        _.defer(callback);
+    };
+    BlobUrlProvider.prototype.clear = function(callback) {
+        var urls = Object.keys(this.urls);
+        urls.forEach(function(url) {
+            URL.revokeObjectURL(url);
+        });
+
+        this.urls = {};
+        this.paths = {};
+
+        _.defer(callback);
+    };
+    BlobUrlProvider.prototype.createURL = function(filename, blob, type, callback) {
+        var self = this;
+
+        // If there's an existing entry for this, remove it.
+        this.remove(filename, function(err) {
+            if(err) {
+                return callback(err);
+            }
+
+            // Now make a new set of cache entries
+            var url = URL.createObjectURL(blob);
+
+            self.urls[filename] = url;
+            self.paths[url] = filename;
+
+            callback(null, url);
+        });
+    };
+    BlobUrlProvider.prototype.remove = function(path, callback) {
+        var self = this;
+        var removed = [];
+
+        // If this is a dir path, look for other paths entries below it
+        Object.keys(self.urls).forEach(function(key) {
+            var url = self.urls[key];
+
+            // The first time a file is written, we won't have
+            // a stale cache entry to clean up.
+            if(!url) {
+                return;
+            }
+
+            // If this filename matches exactly, or is a root path (i.e., other
+            // filenames begin with "<path>/...", remove it. Otherwise just skip.
+            if(key === path || key.indexOf(path + "/") === 0) {
+                removed.push(key);
+
+                delete self.urls[key];
+                delete self.paths[url];
+
+                // Delete the reference from memory
+                URL.revokeObjectURL(url);
+            }
+        });
+
+        _.defer(callback, null, removed);
+    };
+    BlobUrlProvider.prototype.rename = function(oldPath, newPath, callback) {
+        var url = this.urls[oldPath];
+
+        this.urls[newPath] = url;
+        this.paths[url] = newPath;
+        delete this.urls[oldPath];
+
+        _.defer(callback);
+    };
+
+
+    /**
+     * CacheStorageUrlProvider uses CacheStorage and Service Workers in compatible browsers.
+     */
+    function CacheStorageUrlProvider() {
+        UrlProviderBase.call(this);
+
+        // We use cache URLs like https://<origin>/dist/vfs/project/root/filename.ext
+        var prefix = brackets.env === "production" ? "/dist" : "/src";
```

Comment:
Hmm, so this will break if we map `https://mozillathimblelivepreview.net/bramble/staging/dist` to a new origin domain like `https://bramble.net`?